<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<title>查找并修复Hyper应用中的内存泄漏</title>
</head>
<body>
<p><a href="https://zhuanlan.zhihu.com/p/48011141">原文</a></p>
<div class="title-image"><img src="https://pic1.zhimg.com/v2-a76720d67afff53cc37483dc96a096ed_r.jpg" alt=""></div><p>作者对其应用的下一个版本进行负载测试，使用wrk工具发送请求几秒之后，后端内存占用了1.3GB，内存占用的增长速度大约每秒50MB。</p><h2><b>初步调试：</b></h2><p>开始使用heaptrack工具，发现内存只分配了8KB，问题出在了heaptrack工具和jemalloc并不搭，所以换用Rust的system分配器。</p><code lang="rust">
use std::alloc::System;

#[global_allocator]
static A: System = System;</code><p>这样就可以用heaptrack来检测到内存泄漏了。然而还是没有得到什么有用的信息。</p><h2><b>改变策略：</b></h2><p>有一个办法，可以通过逐步删除部分程序来确定内存泄漏的位置。那么有没有这样一个工具来代替开发者手工完成？[C-Reduce](<a href="https://embed.cs.utah.edu/creduce/">https://embed.cs.utah.edu/creduce/</a>)行不行？可惜的是，他们发现C-Reduce无法支持Rust。所以只能想办法徒手实现了一个最小版本的应用示例。</p><code lang="rust">
#![feature(alloc_system)]
#![feature(allocator_api)]
#[global_allocator]
static GLOBAL: std::alloc::System = std::alloc::System;

#[macro_use]
extern crate failure;
extern crate future_utils;
extern crate hyper;
extern crate tokio;

use failure::Error;
use future_utils::{BoxSendFuture, FutureExt};
use hyper::service::Service;
use hyper::Body;
use std::time::Duration;
use tokio::prelude::*;

fn main() {
    let server = hyper::Server::bind(&amp;"127.0.0.1:31337".parse().unwrap())
        .serve(|| Ok::&lt;_, Box&lt;std::error::Error + Send + Sync&gt;&gt;(Server));
    let mut rt = tokio::runtime::Runtime::new().unwrap();
    rt.spawn(server.map_err(move |_err| {}));
    println!("server started");
    ::std::thread::sleep(Duration::from_secs(10));
    println!("shutting down");
    rt.shutdown_now().wait().unwrap();
}

#[derive(Debug, Fail)]
#[fail(display = "myfail")]
pub struct MyFail;

pub struct Server;
impl Service for Server {
    type ReqBody = Body;
    type ResBody = Body;
    type Error = hyper::Error;
    type Future = BoxSendFuture&lt;hyper::Response&lt;Self::ResBody&gt;, Self::Error&gt;;

    fn call(&amp;mut self, _request: hyper::Request&lt;Self::ReqBody&gt;) -&gt; Self::Future {
        Err(Error::from(::MyFail))
            .into_future()
            .or_else(|err: Error| Ok(err.downcast::&lt;::MyFail&gt;().unwrap().into_future()) // removing this gets rid of the leak
            .map(|_| ::hyper::Response::new(Body::empty()))
            .into_send_boxed()
    }
}</code><p>最终发现，是在转换错误类型的时候调用into_future()的时候发生了泄漏。然后他们使用[cargo-geiger](<a href="https://github.com/anderejd/cargo-geiger">anderejd/cargo-geiger</a>)工具，去发现导致这次泄漏的unsafe代码。</p><p><br></p><p>最终他们定位到了问题所在：</p><code lang="rust">
pub(crate) fn downcast&lt;T: Fail&gt;(self) -&gt; Result&lt;T, ErrorImpl&gt; {
        let ret: Option&lt;T&gt; = self.failure().downcast_ref().map(|fail| {
            unsafe {
                // drop the backtrace
                let _ = ptr::read(&amp;self.inner.backtrace as *const Backtrace);
                // read out the fail type
                ptr::read(fail as *const T)
            }
        });
        match ret {
            Some(ret) =&gt; {
                // forget self (backtrace is dropped, failure is moved
                mem::forget(self);
                Ok(ret)
            }
            _ =&gt; Err(self)
        }
    }</code><p>是failure库里的<code class="inline">downcast</code>方法使用了<code class="inline">mem::forget</code>。这就导致了<code class="inline">Error::downcast</code> 并没有把<code class="inline">Box&lt;Inner&lt;Fail&gt;&gt;</code>给drop掉，而仅仅是drop了backtrace。</p><p><br></p><p>之前有一个PR补丁修复了此泄漏漏洞 [failure-pr-262](<a href="https://github.com/rust-lang-nursery/failure/pull/262">Fix a memory leak in downcast by mitsuhiko · Pull Request #262 · rust-lang-nursery/failure</a>)，</p><code lang="rust">// deallocate the box without dropping the inner parts
#[cfg(has_global_alloc)] {
    use std::alloc::{dealloc, Layout};
    unsafe {
        let layout = Layout::for_value(&amp;*self.inner);
        let ptr = Box::into_raw(self.inner);
        dealloc(ptr as *mut u8, layout);
    }
}</code><p>但是在Rust 1.18之前的版本还存在泄漏问题，可见作者用的还是老版本。</p><p>因此，他的同事给Failure提交了一个PR，修复此问题，现在已经被合并到failure中。</p><code lang="rust">pub(crate) fn downcast&lt;T: Fail&gt;(self) -&gt; Result&lt;T, ErrorImpl&gt; {
    if self.failure().__private_get_type_id__() == TypeId::of::&lt;T&gt;() {
        let ErrorImpl { inner } = self;
        let casted = unsafe { Box::from_raw(Box::into_raw(inner) as *mut Inner&lt;T&gt;) };
        let Inner { backtrace:_, failure } = *casted;
        Ok(failure)
    } else {
        Err(self)
    }
}</code><p><br></p><p>[ Read  More ]( <a href="https://blog.1aim.com/2018/10/finding-and-fixing-memory-leaks-in-a-hyper-application-or-how-i-learned-to-stop-worrying-and-love-the-allocator/">https://blog.1aim.com/2018/10/finding-and-fixing-memory-leaks-in-a-hyper-application-or-how-i-learned-to-stop-worrying-and-love-the-allocator/</a> )</p><p></p>
<script async defer="defer" src="https://www.googletagmanager.com/gtag/js?id=UA-7909075-5"></script>
<script>
window.dataLayer = window.dataLayer || [];
function gtag(){ dataLayer.push(arguments); }
gtag('js', new Date());
gtag('config', 'UA-7909075-5');
</script>
<script>
var _hmt = _hmt || [];
</script>
<script async defer="defer" src="https://hm.baidu.com/hm.js?e3d40295e416616ddc21287da9646d31"></script>
</body>
</html>
